### ISSUE TITLE
[Proposal] Better management of subfactories than closure

### ISSUE THREAD

**Issue by ThibaudDauce** [laravel/internals#485]
TLDR;

> We are hardcoding ->create()'s inside of factories that we can call "->make()" on. Therefore, calling "->make()" isn't a sure thing nothing will be persisted to the db...

---

One very common way to use subfactories in Laravel is to make a Closure and call the factory inside:

```php
$factory->define(App\Phone::class, function (Faker\Generator $faker) {
    return [
        'user_id' => function() {
            return factory(App\User::class)->create()->id;
        },
        'phone' => $faker->phoneNumber,
    ];
});
```

The main disadvantage of this approach is that the `create()` method is hard-coded. If I want to use this factory without the `DatabaseMigrations` and a call to `make()` my testing database will be polluted.

```php
factory(App\Phone::class)->make();
$this->assertEquals(0, App\User::count()); // fail, there is one user in my database
```

I think the best API for the users is to use the same key as the name of the relation:

```php
$factory->define(App\Phone::class, function (Faker\Generator $faker) {
    return [
        'user' => factory(App\User::class),
        'phone' => $faker->phoneNumber,
    ];
});
```

```php
$phone = factory(App\Phone::class)->make();
$this->assertEquals(0, App\User::count()); // success, the make function is called on the factory
$this->assertInstanceOf(App\User::class, $phone->user); // success, the relation is populated with `setRelation`
$this->assertNull($phone->user_id); // success, this is the only drawback
```

It will not work well with the `raw()` method. The array will not contain neither a `user` key, nor a `user_id` key.

If it's a wanted feature, *I can submit a pull request*. The `FactoryBuilder` will inspect the model to get the `Relation`, create or make a new model and set the appropriate keys (with `getOwnerKey()`, `getForeignKey()`…). 

PS: I noticed there is no test for the `Factory` and `FactoryBuilder`… :-(

PS2: it could also work with Collection for `HasMany` relationships:

```php
$factory->define(App\User::class, function (Faker\Generator $faker) {
    return [
        'phones' => factory(App\Phone::class, 3),
    ];
});
```

PS3: it could also work with suboverride attributes:

```php
$phone = factory(App\Phone::class)->make([
    'user' => [
        'email' => 'john@example.com',
    ]
]);
$this->assertEquals('john@example.com', $phone->user->email); // success
```

**Comment by Dylan-DPC** [comment]
You are adding more complexity and achieving marginal gains through this approach

**Comment by Garbee** [comment]
Why not structure things where you simply call `make` on things with relations and in your logic do the relating just like you would in your application. Yes, a little extra manual code but it doesn't force any logic into the factories that shouldn't really happen there.

Factories are about creating a given model as much as it can, not completely. If you keep doing this in factories then in your tests you can end up doing a lot more work than necessary to test given cases.

**Comment by Dylan-DPC** [comment]
My only concern is you are using "magic". 

Why not: 
```
return [
        'user_id' => factory(App\User::class),
        'phone' => $faker->phoneNumber,
    ];
```

**Comment by ThibaudDauce** [comment]
@Garbee with this new feature my tests will be much cleaner because I will not need to build my models manually and just use factories.

@Dylan-DPC I think this solution use the same amount of magic than the relations in Laravel. It's the same approach of doing `$user->phone` to access the related model.

**Comment by calebporzio** [comment]
I dig this - seems like a nice idea

**Comment by Dylan-DPC** [comment]
@ThibaudDauce yes but you are using magic where it may not be required. Just adds more complexity

**Comment by skyrpex** [comment]
Just to add more background: you can use a factory state to include relations. For example:

```php
<?php
$factory->define(App\Project::class, function (Faker\Generator $faker) {
    return [
        'name' => $faker->name,
    ];
});
```

```php
<?php
$factory->define(App\Project::class, 'with_user', function (Faker\Generator $faker) {
    return [
        'user_id' => function () { /* ... */ },
    ];
});
```

You can include different states for each relation if you really need to keep them separate.

**Comment by ThibaudDauce** [comment]
@calebporzio You totally understood my point. I'm not entirely sure of using the relation name is the correct solution but the problem of hard-coding `create()` inside a factory with a `make()` method is here. Maybe something else could be possible… I updated my initial post with your sentence in a TLDR :-)

@skyrpex yes, I know it's possible but it doesn't resolve the problem of `create()` inside `make()`…

**Comment by calebporzio** [comment]
@skyrpex - nice!! I like that and will certainly start using that personally.

@ThibaudDauce - I agree it doesn't solve the problem, but it certainly helps to not have phantom "create()"s in a model factory. I think I probably don't have a deep enough understanding of relationships and if / how they exist without persistence - so I'll let those people weigh in.

**Comment by Dylan-DPC** [comment]
@calebporzio 
>  but it certainly helps to not have phantom "create()"s in a model factory. 

any reason for that?

**Comment by calebporzio** [comment]
@Dylan-DPC - ya, it destroys the promise of being able to `->make()` and not have anything persist to the DB
